### PR TITLE
[ENH] Add the ability to convert a DAG to an MAG

### LINKED ISSUE #95: Convert DAG to a MAG
**Is your feature request related to a problem? Please describe.**
I need to be able to convert a DAG to a PAG to generate tests for #93 .

**Describe the solution you'd like**
A function implementing the algorithm to convert DAGs to PAGs as given in Zhang. 2008

### PR DESCRIPTION
**Pull request by aryan26roy**
Closes #95 

Changes proposed in this pull request:

- function to that takes a DAG as an argument and outputs a valid MAG.

The algorithm is described in Zhang et al. 2008 on page number 1877.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
Can you rebase this? The GH files changed tab is showing a lot of files for some reason.

**Comment by aryan26roy** [comment]
Sure.

**Comment by aryan26roy** [comment]
@adam2392 I think it's fine now.

**Comment by aryan26roy** [comment]
@adam2392 I think I figured out how to implement the algorithm. Can you look over the comments and tell me if that is correct or not?

**Review by adam2392 — COMMENTED** [review]
Yeah it looks reasonably correct. Having a few tests, examples that we can refer to visually will be helpful

File: pywhy_graphs/algorithms/generic.py [hunk 1]
```diff
@@ -703,3 +703,56 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
                 return False
 
     return True
+
+
+def dag_to_mag(G, L: Set = None, S: Set = None):
+    """Converts a DAG to a valid MAG.
+    The algorithm is defined in :footcite:`Zhang2008`.
```
[the remark this answers is not in the document]
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:710]: The page numbering is very weird. It's in the thousands and it's not defined as a theorem. I will just put the page number. (although it will look weird)

**Comment by aryan26roy** [comment]
Hey @adam2392 , how do I get/generate valid DAG, MAG pairs? The implementation is complete and I need to test it now.

**Comment by adam2392** [comment]
Nice! 

Perhaps peruse papers or come up with your own and draw them in tetrad

**Comment by aryan26roy** [comment]
@adam2392 I found a pair on the internet and just wanted to confirm, is this a correct DAG and MAG pair? H is a latent variable.

**Comment by adam2392** [comment]
> 

Seems right walking thru each condition of the DAG to MAG algo.

**Comment by aryan26roy** [comment]
I don't see how the bidirected edge from E to R came about. Isn't E present in the ancestral graph of R U S ?

**Comment by adam2392** [comment]
Oh yeah sorry. It should be E -> R right?

See the definition on pg 1877 and the last paragraph on section 2.3. 

Maybe it would help if you explicitly wrote the algorithm into the docstring too.

**Comment by aryan26roy** [comment]
@adam2392 If two nodes are adjacent in a graph, can we say an inducing path exists between the two of them given all the conditions for an inducing path are satisfied?

**Comment by adam2392** [comment]
I believe the definition states this is trivially true. IIRC

**Comment by aryan26roy** [comment]
@adam2392  in the MAG shouldn't A -> E be A - E? My implementation seems to be working correctly.

**Comment by adam2392** [comment]
What's the reasoning?

**Comment by aryan26roy** [comment]
> What's the reasoning?

A is an ancestor of E union S and E is an ancestor of A union S. When both are present in the ancestor sets its a undirected edge, right?

File: pywhy_graphs/algorithms/tests/test_generic.py [hunk 2]
```diff
@@ -348,3 +348,63 @@ def test_valid_mag():
     admg.add_edge("H", "J", admg.undirected_edge_name)
 
     assert not pywhy_graphs.valid_mag(admg)  # there is an undirected edge between H and J
+
+
+def test_dag_to_mag():
```
Comment (adam2392) [inline pywhy_graphs/algorithms/tests/test_generic.py:353]: Can you add comments to the test to make it easier to read?

File: pywhy_graphs/algorithms/generic.py [hunk 3]
```diff
@@ -703,3 +707,77 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
...
+
+    mag = ADMG()
+
+    for elem in adj_nodes:
+
+        temp_list = list(elem)
+        a = set(temp_list[0])
+        b = set(temp_list[1])
+        aus = S.union(a)
+        bus = S.union(b)
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:756]: Perhaps we should use capital lettering to be more clear on variable names.

File: pywhy_graphs/algorithms/generic.py [hunk 4]
```diff
@@ -703,3 +707,77 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
...
+
+    # find the ancestors of B U S (ansB) and A U S (ansA) for each pair of adjacent nodes
+
+    mag = ADMG()
+
+    for elem in adj_nodes:
+
+        temp_list = list(elem)
+        a = set(temp_list[0])
+        b = set(temp_list[1])
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:754]: Isn't adj_nodes just a list of sets of two nodes?

Perhaps it could be simplified to for A,B in adj_nodes?

**Review by adam2392 — CHANGES_REQUESTED** [review]
Nice! I didn't evaluate the tests yet but I left some comments on the readability of the code.

File: pywhy_graphs/algorithms/generic.py (hunk 4 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:754]: This is something I keep forgetting Python can do (even tho it is one of my favourite features)

File: pywhy_graphs/algorithms/generic.py [hunk 5]
```diff
@@ -703,3 +707,74 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
                 return False
 
     return True
+
+
+def dag_to_mag(G, L: Set = None, S: Set = None):
+    """Converts a DAG to a valid MAG.
+    The algorithm is defined in :footcite:`Zhang2008` on page 1877.
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:714]:
```suggestion

    The algorithm is defined in :footcite:`Zhang2008` on page 1877.
```

File: pywhy_graphs/algorithms/generic.py [hunk 6]
```diff
@@ -703,3 +707,74 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
...
+
+    mag = ADMG()
+
+    for A, B in adj_nodes:
+
+        AuS = S.union(A)
+        BuS = S.union(B)
+
+        ansA: Set[str] = set()
+        ansB: Set[str] = set()
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:756]:
```suggestion
        ansA: Set = set()
        ansB: Set = set()
```

Nodes do not necessarily have to be strings

File: pywhy_graphs/algorithms/tests/test_generic.py [hunk 7]
```diff
@@ -348,3 +348,72 @@ def test_valid_mag():
...
```
Comment (adam2392) [inline pywhy_graphs/algorithms/tests/test_generic.py:353]: You should add a test to each mag that it is not almost cyclic and also maximal perhaps even. The graphs are small, so the maximality check should not take long

**Review by adam2392 — COMMENTED** [review]
Okay the code is looking better.

the tests can be improved by explicitly checking for the conditions of the MAG

File: pywhy_graphs/algorithms/tests/test_generic.py (hunk 7 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/tests/test_generic.py:353]: So I added a test to check that no MAG has an ADC. How do I check for maximality? I see the definition in the paper but could you roughly write down the algorithm?

Comment (adam2392) [reply pywhy_graphs/algorithms/tests/test_generic.py:377]: https://www.jmlr.org/papers/volume9/zhang08a/zhang08a.pdf states: "ii. there is no inducing path between any two non-adjacent vertices (maximal)."

Comment (aryan26roy) [reply pywhy_graphs/algorithms/tests/test_generic.py:377]: Ah alright.

Comment (aryan26roy) [reply pywhy_graphs/algorithms/tests/test_generic.py:377]: So I am not passing any L or S variables to the is_maximal function because an MAG cannot have latent or selection variables.

File: pywhy_graphs/algorithms/generic.py [hunk 8]
```diff
@@ -703,3 +708,118 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
...
+    """
+
+    if L is None:
+        L = set()
+
+    if S is None:
+        S = set()
+
+    all_nodes = set(G.nodes)
+    checked = []
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:811]: Don't use list to check presence. It's inefficient. Use set, dict or something else with O(1) efficiency of checking if element is in the container.

Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:811]: I thought of using a data structure with a constant time lookup but none of the mutable data structures are hashable and I need to lookup a pair of nodes.

Comment (adam2392) [reply pywhy_graphs/algorithms/generic.py:811]: Frozenset is hashable

**Review by adam2392 — APPROVED** [review]
LGTM! Thanks @aryan26roy

**Comment by aryan26roy** [comment]
Using frozenset was a better solution!

**Review by adam2392 — CHANGES_REQUESTED** [review]
Actually small code-smell. Otw lgtm

File: pywhy_graphs/algorithms/generic.py [hunk 9]
```diff
@@ -703,3 +708,118 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
...
+    -------
+    mag : Graph
+        The MAG.
+    """
+
+    if L is None:
+        L = set()
+
+    if S is None:
+        S = set()
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:737]: Can you add a test case for graphs when L and S is None?

File: pywhy_graphs/algorithms/generic.py [hunk 10]
```diff
@@ -703,3 +708,118 @@ def valid_mag(G: ADMG, L: set = None, S: set = None):
...
+        nb = set(G.neighbors(source))
+        cur_set = all_nodes - nb
+        cur_set.remove(source)
+        for dest in cur_set:
+            current_pair = frozenset({source, dest})
+            if current_pair not in checked:
+                checked.add(current_pair)
+                out = inducing_path(G, source, dest)
+                if out[0] is True:
+                    return False
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:822]: Perhaps add a different unit-test function to test when maximal is False here? Just take a MAG and add an inducing path somewhere.

File: pywhy_graphs/algorithms/generic.py (hunk 9 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:737]: Done.

File: pywhy_graphs/algorithms/generic.py (hunk 10 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:822]: Done

**Comment by aryan26roy** [comment]
@adam2392 Incorporated all the suggestions you made.
